File: go/weil_go/collections/set.go
package collections

import (
	"encoding/json"
	"fmt"

	"github.com/weilliptic-inc/contract-sdk/go/weil_go/errors"
	"github.com/weilliptic-inc/contract-sdk/go/weil_go/runtime"
)

type __unit__ struct{}

type WeilSet[V comparable] struct {
	InnerMap WeilMap[V, __unit__] `json:"map"`
}

func NewWeilSet[V comparable](stateId WeilId) *WeilSet[V] {
	return &WeilSet[V]{
		InnerMap: WeilMap[V, __unit__]{stateId},
	}
}

// Implement WeilCollection

func (m *WeilSet[V]) BaseStatePath() string {
	return m.InnerMap.BaseStatePath()
}

func (m *WeilSet[V]) StateTreeKey(suffix *V) string {
	serializedSuffix, _ := json.Marshal(suffix)
	return fmt.Sprintf("%s_%s", m.BaseStatePath(), string(serializedSuffix))
}

// Implement WeilSet

func (s *WeilSet[V]) Insert(val *V) {
	runtime.WriteCollectionEntry[V]([]byte(s.StateTreeKey(val)), nil)
}

func (s *WeilSet[V]) Contains(val *V) bool {
	_, err := runtime.ReadCollection[V]([]byte(s.StateTreeKey(val)))
	if err != nil {
		switch err.(type) {
		case *errors.KeyNotFoundInCollectionError:
			return false
		default:
			panic(fmt.Sprintf("error other than `KeyNotFoundInCollection` received while `ReadCollection`: %v", err.Error()))
		}
	}

	return true
}

func (s *WeilSet[V]) Remove(value *V) (bool, error) {
	_, err := runtime.DeleteCollectionEntry[V]([]byte(s.StateTreeKey(value)))
	if err != nil {
		switch err.(type) {
		case *errors.KeyNotFoundInCollectionError:
			return false, nil
		default:
			panic(fmt.Sprintf("error other than `KeyNotFoundInCollection` received while `ReadCollection`: %v", err.Error()))
		}
	}

	return true, nil
}

File: adk/rust/weil_rs/src/collections/set.rs
use super::{map::WeilMap, WeilCollection, WeilId};
use crate::{runtime::Memory, traits::WeilType};
use serde::{Deserialize, Serialize};
use std::{borrow::Borrow, hash::Hash};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct WeilSet<T>(WeilMap<T, ()>);

impl<T> WeilSet<T> {
    pub fn new(id: WeilId) -> Self {
        WeilSet(WeilMap::new(id))
    }
}

impl<T> WeilSet<T>
where
    T: WeilType + Hash + Eq,
{
    pub fn insert(&mut self, value: T) {
        Memory::write_collection::<()>(self.state_tree_key(&value), ())
    }

    pub fn contains<Q>(&self, value: &Q) -> bool
    where
        T: Borrow<Q>,
        Q: Hash + Eq + Serialize,
    {
        Memory::read_collection::<()>(self.state_tree_key(value)).is_some()
    }

    pub fn remove<Q>(&self, value: &Q) -> bool
    where
        T: Borrow<Q>,
        Q: Hash + Eq + Serialize,
    {
        Memory::delete_collection::<()>(self.state_tree_key(value)).is_some()
    }
}

impl<T: WeilType> WeilType for WeilSet<T> {}

impl<'a, T: WeilType, Q: 'a> WeilCollection<'a, Q> for WeilSet<T> {
    type Key = T;

    fn base_state_path(&self) -> WeilId {
        <WeilMap<T, ()> as WeilCollection<'_, Q>>::base_state_path(&self.0)
    }

    fn state_tree_key(&'a self, suffix: &'a Q) -> String
    where
        Q: Serialize,
        <Self as WeilCollection<'a, Q>>::Key: Borrow<Q>,
    {
        <WeilMap<T, ()> as WeilCollection<'_, Q>>::state_tree_key(&self.0, &suffix)
    }
}
